File: kangtaku/partition_to_k_equal_sum_subsets.java
class Solution {
    int[] cache = new int[1 << 16 + 1];
    int allUse;
    int goal;
    int n;
    public boolean canPartitionKSubsets(int[] nums, int k) {
        int sum = 0;
        int i;
        n = nums.length;
        for (i = 0; i < n; i++) {
            sum += nums[i];
        }
        
        if (sum % k != 0) return false;
        
        allUse = (1 << n) - 1;
        goal = sum / k;
        
        return canMake(nums, 0, 0) > 0;
    }
    
    public int canMake(int[] nums, int used, int sum) {
        if (used == allUse) {
            return 1;
        }
        
        if (cache[used] != 0) {
            return cache[used];
        }
        
        int i, n = nums.length;
        int ans = -1;
        for (i = 0; i < n; i++) {
            if ((used & (1 << i)) == 0 && sum + nums[i] <= goal) {
                used |= (1 << i);
                if (sum + nums[i] == goal) {
                    ans = Math.max(ans, canMake(nums, used, 0));
                } else {
                    ans = Math.max(ans, canMake(nums, used, sum + nums[i]));
                }
                if (ans == 1) {
                    return 1;
                }
                used = used & ~(1 << i);
            }
        }
        cache[used] = ans;
        return ans;
    }
}

File: Partition to K Equal Sum Subsets.py
import math

class Solution(object):
    def canPartitionKSubsets(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        nums.sort()
        return self.solve(nums, k, [])

    def solve(self, nums, k, ans):
        if k == 0:
            if ans[0] == ans[1] == ans[2] == ans[3]:
                return True
            return False
        is_found = False
        mid = int(math.ceil(len(nums) / 2))
        for i in range(1, mid + 1):
            if not is_found:
                ans.append(sum(nums[mid - i:mid]))
                is_found = self.solve(nums[:mid - i] + nums[mid:], k - 1, ans)
                ans.pop()

            if not is_found:
                ans.append(sum(nums[mid: mid + i]))
                is_found = self.solve(nums[:mid] + nums[mid + i:], k - 1, ans)
                ans.pop()
        return is_found

if __name__ == '__main__':
    nums = [4, 3, 2, 3, 5, 2, 1]
    k = 4
    sol = Solution()
    print(sol.canPartitionKSubsets(nums, k))